src/tab.ts

```diff
@@ -1,17 +1,163 @@
+import { Tabs } from "./tabs";
+import { sleep, withError } from "./utils";
+
+import clearAllInputs from "./scripts/clearAllInputs";
+
+declare let _tabs: Tabs;
+
 export class Tab {
   active: boolean = false;
   id: number = -1;
+  index: number = -1;
+  windowId: number = -1;
   url: string = "";
   pendingUrl: string = "";
   createdAt: number;
   lastAccessed: number = Date.now();
 
+  //TODO: Improve typehints...
+  connect: CallableFunction;
+  clearAllInputs: CallableFunction;
+  discard: CallableFunction;
+  duplicate: CallableFunction;
+  getLanguage: CallableFunction;
+  getScreenshot: CallableFunction;
+  goBack: CallableFunction;
+  goForward: CallableFunction;
+  move: CallableFunction;
+  reload: CallableFunction;
+  remove: CallableFunction;
+  update: CallableFunction;
+
+  private _removed = false;
+
+  private _withRemoved(cb: CallableFunction) {
+    return (...args: unknown[]) => {
+      if (this._removed) {
+        console.warn("Cannot use operation on closed tab!");
+        return null;
+      }
+      return cb(...args);
+    };
+  }
+
   constructor(tab: chrome.tabs.Tab) {
     Object.assign(this, tab);
     Object.assign(this, {
       [Symbol.toStringTag]: `Tab.${this.urlObj.host || "broken"}`,
     });
+
     this.createdAt = Date.now();
+
+    this.connect = this._withRemoved(this._connect.bind(this));
+    this.clearAllInputs = this._withRemoved(
+      withError(this._clearAllInputs.bind(this)),
+    );
+    this.discard = this._withRemoved(withError(this._discard.bind(this)));
+    this.duplicate = this._withRemoved(withError(this._duplicate.bind(this)));
+    this.goBack = this._withRemoved(withError(this._goBack.bind(this)));
+    this.goForward = this._withRemoved(withError(this._goForward.bind(this)));
+    this.getLanguage = this._withRemoved(withError(this._language.bind(this)));
+    this.getScreenshot = this._withRemoved(
+      withError(this._screenshot.bind(this)),
+    );
+    this.move = this._withRemoved(withError(this._move.bind(this)));
+    this.reload = this._withRemoved(withError(this._reload.bind(this)));
+    this.remove = this._withRemoved(
+      this._withRemoved(withError(this._remove.bind(this))),
+    );
+    this.update = this._withRemoved(withError(this._update.bind(this)));
+
+    this.focus = this._withRemoved(withError(this.focus.bind(this)));
+    this.forceClose = this._withRemoved(withError(this.forceClose.bind(this)));
+  }
+
+  private async _discard(): Promise<Tab> {
+    const rawTab = await getTabs().discard(this.id);
+    _tabs.discard(this.id, rawTab.id!);
+    Object.assign(this, rawTab);
+    return this;
+  }
+
+  public async focus(): Promise<void> {
+    await this._update({ active: true });
+    return;
+  }
+
+  private async _screenshot(options: chrome.tabs.CaptureVisibleTabOptions) {
+    await this.update({ active: true });
+    await sleep(1000);
+    return await getTabs().captureVisibleTab(this.windowId, options);
+  }
```

Comment:
- [ ] Conduct more tests on this one, just to be sure.
- [ ] Look for ways to avoid using `sleep`